README
README
======

IMPORTANT:
==========
In the accountview you have the possibility to copy the decrytped password to your clipboard.
If you use the firefox/mozilla browser you have to change your firefox security settings to use this feature.
Therefor add the fallowing line to your firefox/mozilla configfile (prefs.js). This file is located under your firefox user profile directory.
	user_pref("signed.applets.codebase_principal_support", true);
Another way is to change this setting directly over your browser interface.
Therefor call the URL "about:config", press the right button and insert a new boolean value
with name "signed.applets.codebase_principal_support" and the value "true".

It also make sense to protect the the phpPassManager pages with "htaccess" and use SSL to access the pages!



Requirements:
=============
- Apache
- PHP
- MySQL
- PHP mcrypt module



Install:
========

1: extract package
Extract the file phppassmanager-x.yy.zz.tgz in your webserver root directory.
On your shell you can do this with: tar -xfvz phppassmanager-x.yy.zz.tgz

2: create database
Create a MySQL database or use a already existing one.
For example on your shell you can create a databse this with command: echo "create database phppassmanager" | mysql -u username --password=password

3: create tables
Use the file "install/tables.sql" to create the tables.
You can do this for example with phpMyAdmin or on your shell: mysql -u username --password=password phppassmanager < tables.sql

4: create a db user
create a db user or use a existing one

5: edit config-<sitename>.php
<sitename> == to the first section of the vhosts fqdn.
Edit the file "config-<sitename>.php"
Fill in there your dbusername, dbpassword an the database you want to use

6: ready
Call the website with http://<sitename>.<domainname>



changelog.txt
2006-12-23
==========
release 0.90.20
- change: changed the random number generator for the crypt function from DEV_RADNOM to DEV_URANDOM; gives extreme performance boost.

2006-11-02
==========
release 0.90.10
- bugfix: fixed a bug in the database update/create scripts. Bug causes that the config parameter "on decode check password" can't be saved.

2006-10-30
==========
release 0.90.00
- change: on decode the masterpassword is checked whether it is right or not

2006-10-22
==========
release 0.89.85
- change: the searchkeyword is now also searched in the URL field
- change: on the account-add, edit, edit-pass and view-page the first textfield get automatically the focus
- change: highlighting the table rows on mouseover
- change: show accesskey help an each page

2006-10-16
==========
release 0.89.75
- bugfix: fixed a bug that sometimes occurs when a account is saved [clsPpmCrypt.php, pp_account_save.php]
- change: on activated search (group and/or text), the info text "filter on" is shown in red, othwerwise in green
- change: same changes done on the page design / color theme.
- change: all main function now accessable with accesskeys

2006-10-14
==========
release 0.89.55
- change: added a function to check the communication with the crypt module [clsPpmCrypt.php]

2006-10-12
==========
release 0.89.45
- bugfix: fixed the bug that a account can't be saved with Internet Explorer

2006-10-11
==========
release 0.89.35
- change: some internal code cleanup done; changes made on the save page and the initial-vector creation for the crypt function
- change: changed the config parameter "accounts per page" to show more values (25, 30, 50), and the possibility to show all accounts in the mainview

2006-10-09
==========
release 0.89.15
- bugfix: fixed the bug that the url field not shown in mainview; bug first appeared in version 0.89.10 [index.php]

2006-10-09
==========
release 0.89.10
- bugfix: bugs in the page handling fixed. Mainly variable transfer between the pages to keep the filter settings when returning to mainpage.
- change: some internal code cleanup done (variable-processing, -names, handling and transfer)

2006-10-08
==========
release 0.89.00
- new:    divided the mainview into pages. How many accounts shown per page can be configured over a config value. [index.php, ppm_config.php]

2006-10-06
==========
release 0.88.20
- bugfix: another bug in the table create statements fixed. Now also will with MySQL 3.x.

2006-10-05
==========
release 0.88.10
- bugfix: a bug in the tables create statements fixed. They now also will work with MySQL 4.x and not only with MySQL 5.x.

2006-10-04
==========
release 0.88.00
- change: some pagedesign changes done on the config- and mainpage [index.php, ppm_config.php]
- new:    added the possibility to sort the accounts (ascending/descending) by name, group, login or url [index.php]

2006-09-30
==========
release 0.87.00
- new:    config parameter to use the accountname as link to the account view page [index.php, ppm_config]
- new:    added the field "URL" to the account 
- change: adjusted the input textfields to larger size (add and edit) [ppm_account_edit.php, ppm_account_edit_pass]]
- change: changed the table with an the startpage [index.php]
- new:    added a date field to show the create date of the account [ppm_account_view.php]
- new:    added a date field to show the modify date of the account [ppm_account_view.php]
- new:    added a counter to show how often the account has been viewed [ppm_account_view.php]
- new:    added a counter to show how often the account has been decrypted [ppm_account_view.php]
- change: moved the account management function to a class [clsPpmAccount.php]
- change: files "install/tables.sql" updated

2006-09-29
==========
release 0.82.00
- new:    config parameter "show encrypted password" added
- change: function "random pass" provided with a popup the this function will be implemented later [ppm_account_add.php]

2006-09-25
==========
release 0.81.10
- change: updated readme.txt to explain how to use the javascript function "copy to clipboard" with firefox/mozilla browers

2006-09-24
==========
release 0.81.00
- change: completly rewriten the save function [ppm_account_save.php]
- change: moved the crypt-functions to a own class [clsPpmCrypt]
- bugfix: sometimes the create/edit-pass functions causes a database error [ppm_account_save.php]

2006-09-22
==========
release 0.80.10
- first release of phpPassManager


ppm_account_view.php
<?php

        require_once("constants.inc");
        require_once(CONFIG_FILE);
        require_once("functions/db.php");
        require_once("functions/pagedesign.php");
        require_once("classes/clsPpmCrypt.php");
        require_once("classes/clsPpmConfig.php");
        require_once("classes/clsPpmAccount.php");

        StartRenderTimer();

	$clsAccount = new clsAccount;
	$objCrypt = new clsPpmCrypt;

        // database connect
        $dbh = OpenDatabase($DBHost, $DbUser, $DbPass, $DbName);

	// get vars
	$intAccountId = $_POST["accountid"];
	$intDecode = $_POST["decode"];
	$intDelete = $_POST["delete"];
	$strMasterpass = $_POST["masterpass"];
	$strSortKey = $_POST["skey"];
	$strSortOrder = $_POST["sorder"];
	$strGroup = $_POST["group"];
	$strSearch = $_POST["search"];
	$intPage = $_POST["page"];


	$intAccountId_get = $_GET["id"];

	if ( $intAccountId_get != "" ) {
		$intAccountId = $intAccountId_get;
	}

	if ( $intDecode == 1 ) {
        	$clsAccount->incrementDecryptCounter($intAccountId);
	} else {
        	$clsAccount->incrementViewCounter($intAccountId);
	}

	$clsAccount->getAccount($intAccountId);

?>
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<HTML>
<HEAD>
	<TITLE><?php echo (HTML_TITLE) ?></TITLE>
        <LINK REL="stylesheet" HREF="styles.css" TYPE="text/css">
	<SCRIPT SRC="javascript/clipboard.js" LANGUAGE="javascript" TYPE="text/javascript"></SCRIPT>
</HEAD>
<BODY ONLOAD="javascript:document.showaccount.masterpass.focus()">

<?php

        PrintHeader();

?>

<DIV ALIGN="center">

<?php

	if ( $intDelete == 1 ) {
		print "<h2>delete account</h2>";
	} else {
		print "<h2>show account</h2>";
	}

	if ( $intDecode == 1 ) {
		if ( $strMasterpass == "" && COMMON_MASTER_PASS == false) {

?>

<H1>to decode you have to use a 'Masterpassword'!</H1>
<P>click <I><A HREF="javascript:history.back()">back</A></I> to go back</P>
</DIV>

<?php
			die;
		}
	}

//	$query = "SELECT intAccountId, vacName, vacGroupName, vacLogin, vacUrl, vacPassword, vacInitialValue, txtNotice, intCountView, intCountDecrypt, datAdded, datChanged FROM accounts LEFT JOIN groups ON accounts.intGroupFid=groups.intGroupId WHERE intAccountId = $intAccountIdid";
//	$result = mysql_query($query) or die("query failed: " . mysql_error());

//	$row = mysql_fetch_array($result, MYSQL_ASSOC);

	if ( $intDecode == 1 ) {
		$strDecrypted = $objCrypt->decrypt($clsAccount->strPassword, $strMasterpass, $clsAccount->strInitialVector);
	}

?>

<TABLE WIDTH="80%">
	<TR CLASS="odd">
		<TD ALIGN="right">Name:</TD>
		<TD><?php echo ($clsAccount->strName) ?></TD>
	</TR>
	<TR CLASS="even">
		<TD WIDTH="25%" ALIGN="right">Group:</TD>
		<TD><?php echo ($clsAccount->strGroupName) ?></TD>
	</TR>
	<TR CLASS="odd">
		<TD WIDTH="25%" ALIGN="right">Login:</TD>
		<TD><?php echo ($clsAccount->strLogin) ?></TD>
	</TR>
        <TR CLASS="even">
                <TD WIDTH="25%" ALIGN="right">URL:</TD>
                <TD><?php echo ("<A HREF=\"". $clsAccount->strUrl . "\" target=\"_blank\">" . $clsAccount->strUrl . "</A>") ?></TD>
        </TR>
	<TR CLASS="odd">
		<TD WIDTH="25%" ALIGN="right">Password:</TD>
		<TD>
			<TABLE CLASS="altTable" WIDTH="100%"><TR><TD>
<?php

	 if ( $intDecode == 1 ) {

		$clsConfig = new clsConfig;

		$blnNoMd5 = FALSE;
		if ( $clsConfig->getConfigValue("md5_pass") == "TRUE" ) {
			if ($clsAccount->strMd5Password != "0") {
				if (md5($strDecrypted) == $clsAccount->strMd5Password) {	
					$blnDecryptCheck = TRUE;
					if ( $clsConfig->getConfigValue("password_show") == "FALSE" ) {
						$strPasswordOutput = "password not shown - can only copy to clipboard";
					} else {
			 			$strPasswordOutput = $strDecrypted;
					}
				} else {
					$strPasswordOutput = "<B CLASS=\"altTextRed\">Wrong Masterpassword!</B>";
					$blnDecryptCheck = FALSE;
				}
			} else {
				$blnDecryptCheck = TRUE;
				$blnNoMd5 = TRUE;
				$strPasswordOutput = $strDecrypted . " <B CLASS=\"altTextRed\">**</B>";
			}
		} else {
			if ( $clsConfig->getConfigValue("password_show") == "FALSE" ) {
				$blnDecryptCheck = TRUE;
				$strPasswordOutput = "password not shown - can only copy to clipboard";
			} else {
				$blnDecryptCheck = TRUE;
				$strPasswordOutput = $strDecrypted;
			}
		}

		echo ($strPasswordOutput);

	 } else {
		 echo ("not available - decrypted");} ?>

			<FORM ACTION="">
<?php

	if ( $intDecode == 1 AND $blnDecryptCheck == TRUE ) {

?>

			</TD><TD ALIGN="right">
				<INPUT TYPE="button" NAME="clipboard" VALUE="copy to Clipboard" CLASS="altButtonFormat" ONCLICK="copy_clip('<?php echo ($strDecrypted) ?>')">

<?php

	}

?>

			</TD></TR></TABLE>
			</FORM>
		</TD>
	</TR>
	<TR CLASS="even">
		<TD WIDTH="25%" ALIGN="right">Notice:</TD>
		<TD><?php echo ($clsAccount->strNotice) ?></TD>
	</TR>
        <TR CLASS="odd">
                <TD WIDTH="25%" ALIGN="right">views:</TD>
                <TD><?php echo ($clsAccount->intCountView) ?></TD>
        </TR>
        <TR CLASS="even">
                <TD WIDTH="25%" ALIGN="right">decrypts:</TD>
                <TD><?php echo ($clsAccount->intCountDecrypt) ?></TD>
        </TR>
        <TR CLASS="odd">
                <TD WIDTH="25%" ALIGN="right">added:</TD>
                <TD><?php echo ($clsAccount->strDatAdded) ?></TD>
        </TR>
        <TR CLASS="even">
                <TD WIDTH="25%" ALIGN="right">changed:</TD>
                <TD><?php echo ($clsAccount->strDatChanged) ?></TD>
        </TR>
</TABLE>

<?php

	if ( $blnNoMd5 == TRUE ) {

?>

	<TABLE WIDTH="80%" CLASS="altTable">
		<TR>
			<TD ALIGN="right">
				<B CLASS="altTextRed">
					** can't verify password, because internal checksum not available.<BR>
					To write checksum press the button.
				</B>
			</TD>
			<TD>
		                        <FORM ACTION="ppm_account_md5_save.php" METHOD="post">
		                                <INPUT TYPE="hidden" NAME="accountid" VALUE="<?php echo ($intAccountId) ?>">
                		                <INPUT TYPE="hidden" NAME="md5sum" VALUE="<?php echo (md5($strDecrypted)) ?>">
		                                <INPUT TYPE="submit" VALUE="write checksum" CLASS="altButtonFormat" ACCESSKEY="d">
					</FORM>
			</TD>
		</TR>
	</TABLE>

<?php

	}

?>

<BR><BR>

<TABLE WIDTH="80%">
	<TR ALIGN="center">

<?php

	if ( $intDelete == 0 AND $intDecode == 0 ) {

?>

		<TD WIDTH="50%">
			<FORM ACTION="ppm_account_view.php" METHOD="post" NAME="showaccount">
				<INPUT TYPE="hidden" NAME="accountid" VALUE="<?php echo ($clsAccount->intAccountId) ?>">
				<INPUT TYPE="hidden" NAME="decode" VALUE="1">
				<INPUT TYPE="hidden" NAME="skey" VALUE="<?php echo ($strSortKey) ?>">
				<INPUT TYPE="hidden" NAME="sorder" VALUE="<?php echo ($strSortOrder) ?>">
				<INPUT TYPE="hidden" NAME="group" VALUE="<?php echo ($strGroup) ?>">
				<INPUT TYPE="hidden" NAME="search" VALUE="<?php echo ($strSearch) ?>">
				<INPUT TYPE="hidden" NAME="page" VALUE="<?php echo ($intPage) ?>">
				<INPUT TYPE="submit" VALUE="decode" CLASS="altButtonFormat" ACCESSKEY="d">
				<?php if (COMMON_MASTER_PASS == false) { ?>
        Masterpassword:
				<INPUT TYPE="password" NAME="masterpass">
        <?php } ?>
			</FORM>
		</TD>

<?php

	}
	if ( $intDelete == 1 ) {

?>

		<TD WIDTH="50%">
			<FORM ACTION="ppm_account_save.php" METHOD="post">
				<INPUT TYPE="hidden" NAME="accountid" VALUE="<?php echo ($clsAccount->intAccountId) ?>">
				<INPUT TYPE="hidden" NAME="savetyp" VALUE="3">
				<INPUT TYPE="hidden" NAME="skey" VALUE="<?php echo ($strSortKey) ?>">
				<INPUT TYPE="hidden" NAME="sorder" VALUE="<?php echo ($strSortOrder) ?>">
				<INPUT TYPE="hidden" NAME="group" VALUE="<?php echo ($strGroup) ?>">
				<INPUT TYPE="hidden" NAME="search" VALUE="<?php echo ($strSearch) ?>">
				<INPUT TYPE="hidden" NAME="page" VALUE="<?php echo ($intPage) ?>">
				<INPUT TYPE="submit" VALUE="delete" CLASS="altButtonFormat" ACCESSKEY="d">
			</FORM>
		</TD>

<?php

	}

?>

		<TD WIDTH="50%">
			<FORM ACTION="index.php" METHOD="get">
				<INPUT TYPE="hidden" NAME="skey" VALUE="<?php echo ($strSortKey) ?>">
				<INPUT TYPE="hidden" NAME="sorder" VALUE="<?php echo ($strSortOrder) ?>">
				<INPUT TYPE="hidden" NAME="group" VALUE="<?php echo ($strGroup) ?>">
				<INPUT TYPE="hidden" NAME="search" VALUE="<?php echo ($strSearch) ?>">
				<INPUT TYPE="hidden" NAME="page" VALUE="<?php echo ($intPage) ?>">
				<INPUT TYPE="submit" VALUE="back" CLASS="altButtonFormat" ACCESSKEY="b">
			</FORM>
		</TD>
	</TR>
</TABLE>

<BR><BR>
<TABLE CLASS="altTableHelp">
        <TR>
                <TD>

<?php

	if ( $intDelete == 0 AND $intDecode == 0 ) {

?>

                        <B>alt&d:</B> decode
                        <B>alt&b:</B> back

<?php

	} elseif ( $intDelete == 1 ) {

?>

                        <B>alt&d:</B> delete
                        <B>alt&b:</B> back

<?php

	} else {

?>

                        <B>alt&b:</B> back

<?php

	}

?>

                </TD>
        </TR>
</TABLE>

<?php

//	mysql_free_result($result);

	mysql_close($dbh);

        PrintFooter();

        $intRenderTime=StopRendertimer();
        PrintRenderTime ($intRenderTime);

?>

</DIV>

</BODY>
</HTML>
